fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -558,27 +563,55 @@ private void increaseLimit() {
         // Note: This runs all of the updates in serial in order to not invoke a race condition
         // in the rank code that was causing incorrect results. If everything were thread safe,
         // a larger pipeline size would be possible.
-        return cursor.forEachResultAsync(result -> {
+
+        final AtomicReference<RecordCursorResult<FDBStoredRecord<Message>>> holder = new AtomicReference<>(RecordCursorResult.exhausted());
+        final FDBRecordContext context = store.getContext();
+        return AsyncUtil.whileTrue(() -> cursor.onNext().thenCompose(result -> {
+
+            if (!result.hasNext()) {
+                // end of the cursor list
+                if (timer != null) {
+                    timer.increment(FDBStoreTimer.Counts.ONLINE_INDEX_BUILDER_RANGE_BY_COUNT);
+                }
+                holder.set(result);
+                return AsyncUtil.READY_FALSE;
+            }
+
+            // here: implement forEachResultAsync
             final FDBStoredRecord<Message> rec = result.get();
             empty.set(false);
             if (timer != null) {
                 timer.increment(FDBStoreTimer.Counts.ONLINE_INDEX_BUILDER_RECORDS_SCANNED);
             }
             recordsScannedCounter.incrementAndGet();
-            if (recordTypes.contains(rec.getRecordType())) {
-                if (timer != null) {
-                    timer.increment(FDBStoreTimer.Counts.ONLINE_INDEX_BUILDER_RECORDS_INDEXED);
-                }
-                if (syntheticPlan == null) {
-                    return maintainer.update(null, rec);
-                } else {
-                    // Pipeline size is 1, since not all maintainers are thread-safe.
-                    return syntheticPlan.execute(store, rec).forEachAsync(syntheticRecord -> maintainer.update(null, syntheticRecord), 1);
-                }
-            } else {
-                return AsyncUtil.DONE;
+            if (!recordTypes.contains(rec.getRecordType())) {
+                // This record is not our type, swipe left
+                return AsyncUtil.READY_TRUE;
             }
-        }).thenCompose(noNextResult -> {
+            // here: add this index to the transaction
+            if (timer != null) {
+                timer.increment(FDBStoreTimer.Counts.ONLINE_INDEX_BUILDER_RECORDS_INDEXED);
+            }
+            return (syntheticPlan == null ?
+                    maintainer.update(null, rec) :
+                    syntheticPlan.execute(store, rec)
+                            .forEachAsync(syntheticRecord -> maintainer.update(null, syntheticRecord), 1)
+                    ).thenCompose(vignore ->
+                    context.getApproximateTransactionSize().thenCompose(size -> {
```

Comment:
I think this can be `.thenApply` instead of `.thenCompose`, and then it would return `false` and `true` instead of `READY_FALSE` and `READY_TRUE`.

Reply:
I've tried that and it didn't work. I assume that the expectations is for a boolean completable future is not happen with the thenApply.

Reply:
What _exactly_ did you try? It should work if you use `.thenApply` and return vanilla booleans (not CompletableFutures)?